### PR TITLE
Initial Browser test

### PR DESCRIPTION
**Pull request by litzenberger**
Adding browser benchmarks

screen shot 2018-12-21 at 12 49 32 am

### CODE REVIEW

File: tests/lib/create-node.js [hunk 1]
```diff
@@ -49,6 +55,21 @@ const CreateNodeJs = async (config, init, IPFS, count) => {
...
 }
 
+const CreateBrowser = async (config, init, IPFS, count) => {
+  return new Promise((resolve, reject) => {
+
+  // create deamon
+  let daemon
+  let client
+  const f = IPFSFactory.create()
+  f.spawn({ initOptions: { bits: 1024 } }, (err, _ipfsd) => {
```
Comment (mcollina) [inline tests/lib/create-node.js]: Can you promisify spawn instead?

File: tests/lib/create-node.js [hunk 2]
```diff
@@ -35,7 +38,10 @@ const CreateNodeJs = async (config, init, IPFS, count) => {
         .toString()
         .substring(2, 8)}`,
       config: config || defaultConfig[count],
-      init: init || { privateKey: privateKey[count].privKey }
+      init: init || { privateKey: privateKey[count].privKey },
+      EXPERIMENTAL: {
+        pubsub: true
+      }
```
Comment (mcollina) [inline tests/lib/create-node.js]: Is this needed?

Comment (litzenberger) [reply tests/lib/create-node.js]: no

File: tests/lib/create-node.js [hunk 3]
```diff
@@ -6,89 +6,109 @@ const defaultConfig = require('../config/default-config.json')
...
+const fsMakeDir = util.promisify(fs.mkdir)
+const fsWriteFile = util.promisify(fs.writeFile)
+const rimrafPromise = util.promisify(rimraf)
 const privateKey = require('../config/private-key.json')
 const os = require('os')
 const conf = { tmpPath: os.tmpdir() }
 const { repoPath } = require('../package.json').config
+const ipfsClient = require('ipfs-http-client')
+const IPFSFactory = require('ipfsd-ctl')
+
```
Comment (mcollina) [inline tests/lib/create-node.js]: this space is not needed

File: tests/lib/create-node.js [hunk 4]
```diff
@@ -6,89 +6,109 @@ const defaultConfig = require('../config/default-config.json')
...
-    })
+  let init = spawn('ipfs', ['init'], { env: Object.assign(process.env, { IPFS_PATH: path }) })
+  init.stdout.on('data', (data) => {
+    console.log(`stdout: ${data}`)
+  })
+  init.stderr.on('data', (errorData) => {
+    console.error(`${errorData}`)
+  })
+  init.on('close', (code, signal) => {
+    console.log('Repo iinitialized')
```
Comment (mcollina) [inline tests/lib/create-node.js]:
```suggestion
    console.log('Repo initialized')
```

File: tests/lib/create-node.js [hunk 5]
```diff
@@ -6,89 +6,109 @@ const defaultConfig = require('../config/default-config.json')
...
-    init.on('close', (code, signal) => {
-      console.error('Repo iinitialized')
-      resolve(init)
-    })
+  let init = spawn('ipfs', ['init'], { env: Object.assign(process.env, { IPFS_PATH: path }) })
+  init.stdout.on('data', (data) => {
+    console.log(`stdout: ${data}`)
+  })
+  init.stderr.on('data', (errorData) => {
+    console.error(`${errorData}`)
```
Comment (mcollina) [inline tests/lib/create-node.js]:
```suggestion
    console.error(`stderr: ${errorData}`)
```

File: tests/lib/create-node.js [hunk 6]
```diff
@@ -6,89 +6,109 @@ const defaultConfig = require('../config/default-config.json')
...
+      .toString()
+      .substring(2, 8)}`,
+    config: config || defaultConfig[count],
+    init: init || { privateKey: privateKey[count].privKey }
+  })
+  node.on('ready', () => {
+    console.log('Node ready')
+  })
+  node.on('error', (err) => {
+    console.error(`${err}`)
```
Comment (mcollina) [inline tests/lib/create-node.js]:
```suggestion
    console.error(err)
```

File: tests/lib/create-node.js [hunk 7]
```diff
@@ -6,89 +6,109 @@ const defaultConfig = require('../config/default-config.json')
...
+    }
+    if (data.includes('Daemon is ready')) {
+      peer.emit('done')
+    }
+  })
+  peer.stderr.on('data', (data) => {
+    console.error(`${data}`)
+  })
+  peer.on('close', (code, signal) => {
+    console.error('Daemon is stopped')
```
Comment (mcollina) [inline tests/lib/create-node.js]:
```suggestion
    console.error(`Daemon exited with code: ${code}`)
```

**Comment by litzenberger** [comment]
@alanshaw I replaced the HTTP client and get this error:
`Error: WebSocket Servers can not be created in the browser!`


`

File: tests/browser/.gitignore [hunk 8]
```diff
@@ -0,0 +1,23 @@
+# See https://help.github.com/articles/ignoring-files/ for more about ignoring files.
```
Comment (mcollina) [inline tests/browser/.gitignore]: this file should be merged with the top-level `.gitignore`

File: tests/browser/README.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+Instead, it will copy all the configuration files and the transitive dependencies (Webpack, Babel, ESLint, etc) right into your project so you have full control over them. All of the commands except `eject` will still work, but they will point to the copied scripts so you can tweak them. At this point you’re on your own.
+
+You don’t have to ever use `eject`. The curated feature set is suitable for small and middle deployments, and you shouldn’t feel obligated to use this feature. However we understand that this tool wouldn’t be useful if you couldn’t customize it when you are ready for it.
+
+## Learn More
+
+You can learn more in the [Create React App documentation](https://facebook.github.io/create-react-app/docs/getting-started).
+
+To learn React, check out the [React documentation](https://reactjs.org/).
```
Comment (mcollina) [inline tests/browser/README.md]: Can you please change this readme file with a bit of instructions on what this is?

File: tests/browser/public/index.html [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      If you open it directly in the browser, you will see an empty page.
+
+      You can add webfonts, meta tags, or analytics to this file.
+      The build step will place the bundled scripts into the <body> tag.
+
+      To begin the development, run `npm start` or `yarn start`.
+      To create a production bundle, use `npm run build` or `yarn build`.
+    -->
+  </body>
+</html>
```
Comment (mcollina) [inline tests/browser/public/index.html]: There are a lot of comments in here, can you remove them?

File: tests/browser/public/manifest.json [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+      "src": "favicon.ico",
+      "sizes": "64x64 32x32 24x24 16x16",
+      "type": "image/x-icon"
+    }
+  ],
+  "start_url": ".",
+  "display": "standalone",
+  "theme_color": "#000000",
+  "background_color": "#ffffff"
+}
```
Comment (mcollina) [inline tests/browser/public/manifest.json]: This file looks like scaffold. Is it needed? If not, can we remove?

File: tests/browser/src/App.js [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+    let node
+
+    create()
+
+    function create () {
+      // Create the IPFS node instance
+      const start = hrtime()
+
+
+      node = new IPFS({ repo: String(Math.random() + Date.now()) })
```
Comment (mcollina) [inline tests/browser/src/App.js]: a uuid is probably better than this type of random generation

File: tests/browser/src/App.js [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        if (err) {
+          throw err
+        }
+        self.setState({
+          id: res.id,
+          version: res.agentVersion,
+          protocol_version: res.protocolVersion,
+          time_s: delta[0],
+          time_ms: delta[1]
+        })
```
Comment (mcollina) [inline tests/browser/src/App.js]: How is this read from the file in `test/`? Can you add a comment/docs on how this information is moved?

File: tests/browser/src/App.test.js [hunk 14]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import App from './App';
+
+it('renders without crashing', () => {
+  const div = document.createElement('div');
+  ReactDOM.render(<App />, div);
+  ReactDOM.unmountComponentAtNode(div);
+});
```
Comment (mcollina) [inline tests/browser/src/App.test.js]: I think we can remove the tests and their dependencies

File: tests/browser/src/index.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import React from 'react';
+import ReactDOM from 'react-dom';
+import './index.css';
+import App from './App';
+import * as serviceWorker from './serviceWorker';
```
Comment (mcollina) [inline tests/browser/src/index.js]: I don't think we need service workers.

File: tests/browser/src/logo.svg [hunk 16]
```diff
@@ -0,0 +1,7 @@
...
+        <circle cx="420.9" cy="296.5" r="45.7"/>
+        <path d="M520.5 78.1z"/>
+    </g>
+</svg>
```
Comment (mcollina) [inline tests/browser/src/logo.svg]: I don't think we need this file.

File: tests/browser/src/serviceWorker.js [hunk 17]
```diff
@@ -0,0 +1,135 @@
+// This optional code is used to register a service worker.
```
Comment (mcollina) [inline tests/browser/src/serviceWorker.js]: I don't think we need this file

File: tests/init-node.browser.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
+'use strict'
+
+const { build } = require('./schema/results')
+const run = require('./lib/runner')
+
+async function initializeNodeBrowser (node, name, warmup, fileSet, version) {
+  const page = node[0].page
+  await page.waitFor(3000)
```
Comment (mcollina) [inline tests/init-node.browser.js]: what is this timeout? Can we wait for an event instead?

**Comment by litzenberger** [comment]
@mcollina I'll update the dashboard in a new PR.

**Comment by litzenberger** [comment]
I was planning on doing the runner also, its part of the dashboard PR. I'll add a few of the outstanding tests to it to get the dashboard up-to-date.
cc: @Elexy

File: tests/init-node.browser.js (hunk 18 above)
Comment (litzenberger) [reply tests/init-node.browser.js]: yes, will be making this a little better from an app standpoint using buttons to start each test.  coming soon.

Comment (litzenberger) [reply tests/init-node.browser.js]: This has been fixed in this PR: https://github.com/ipfs/benchmarks/pull/94

**Pull request merged**